### PR TITLE
Validate Structured Fields, with adjustments

### PR DESCRIPTION
**Pull request by mnot**
This annotates examples with `http-message` so they can be validated; I corrected the errors I saw.

One outstanding issue is that you're wrapping long binary content across multiple lines, and HTTP line folding rules means that doing so inserts whitespace, which isn't allowed in structured binary content.

Two possible solutions:

1. Make a note near the top that it's wrapping these long lines for illustrative purposes only, and that they're not valid; I can adjust the tooling to make it validate properly despite that
2. Change structured binary content's definition to allow whitespace

(2) is unlikely, because that document is in the RFC Editor queue, but not impossible.

### CODE REVIEW

**Comment by Lukasa** [comment]
Let’s not do (2), especially not for this reason.

**Comment by mnot** [comment]
@Lukasa I hear you, but this is turning out to be a nasty issue -- a spec can use folding where whitespace is necessary; consider:

~~~ http-message
Signature-Input: sig1=(*request-target *created host date
     cache-control x-empty-header x-example); keyid="test-key-a";
     alg=hs2019; created=1402170695; expires=1402170995
 Signature: sig1=:K2qGT5srn2OGbOIDzQ6kYT+ruaycnDAAUpKv+ePFfD0RAxn/1BUe
     Zx/Kdrq32DrfakQ6bPsvB9aqZqognNT6be4olHROIkeV879RrsrObury8L9SCEibe
    oHyqU/yCjphSmEdd7WD+zrchK57quskKwRefy2iEC5S2uAH0EPyOZKWlvbKmKu5q4
    CaB8X/I5/+HLZLGvDiezqi6/7p2Gngf5hwZ0lSdy39vyNMaaAT0tKo6nuVw0S1MVg
    1Q7MpWYZs0soHjttq0uLIA3DIbQfLiIvK6/l0BdWTU7+2uQj7lBkQAsFZHoA96ZZg
    FquQrXRlmYOh+Hx5D9fJkXcXe5tmAg==:
~~~

Here, the inner list in `Signature-Input` requires folding to insert a whitespace, otherwise you lose delimitation; but `Signature` requires it to be withheld. 

Addressing this with (1) means that you'd need to specify different whitespace folding rules for different headers, which is beyond ugly. It's also likely to cause bugs from people who only read examples...

Also, base64 implementations already ignore whitespace; e.g., 

~~~ python
>>> base64.encodestring(b"foo bar baz")
b'Zm9vIGJhciBiYXo=\n'
>>> base64.decodestring(b'Zm9vIGJhciBiYXo=\n')
b'foo bar baz'
>>> base64.decodestring(b'Zm9vIGJhciBiY Xo=\n')
b'foo bar baz'
>>> base64.decodestring(b'Zm9vIGJhciBiY    Xo=\n')
b'foo bar baz'
~~~

... which means that all we need to do is make a small change to allow whitespace [here](https://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#name-byte-sequences) and [here](https://httpwg.org/http-extensions/draft-ietf-httpbis-header-structure.html#name-parsing-a-byte-sequence).

I know it's 11:59:59, but I'm starting to think this change is worth it. Will start a thread on the mailing list.

**Comment by Lukasa** [comment]
Is that a bug that actually exists? Do people read examples and start adding newlines to their HTTP header fields? This seems surprising to me, though it’s within the realms of possibility.

It seems fine to me, in the abstract, to say that we can have folding in specs that are forbidden by the grammar of the field. We already do this in a number of other header fields, so this wouldn’t be the first time. But I’m open to discussion on-list.

**Comment by phluid61** [comment]
Brainstorming: What if it doesn't look like HTTP folding?

~~~
Signature-Input: sig1=(*request-target *created host date
    cache-control x-empty-header x-example); keyid="test-key-a";
    alg=hs2019; created=1402170695; expires=1402170995
Signature: sig1=:K2qGT5srn2OGbOIDzQ6kYT+ruaycnDAAUpKv+ePFfD0RAxn/1BUe
Zx/Kdrq32DrfakQ6bPsvB9aqZqognNT6be4olHROIkeV879RrsrObury8L9SCEibeoHyq
U/yCjphSmEdd7WD+zrchK57quskKwRefy2iEC5S2uAH0EPyOZKWlvbKmKu5q4CaB8X/I5
/+HLZLGvDiezqi6/7p2Gngf5hwZ0lSdy39vyNMaaAT0tKo6nuVw0S1MVg1Q7MpWYZs0so
Hjttq0uLIA3DIbQfLiIvK6/l0BdWTU7+2uQj7lBkQAsFZHoA96ZZgFquQrXRlmYOh+Hx5
D9fJkXcXe5tmAg==:
~~~

And that's without restorting to other conventions, like `\\\n` line continuations.

**Comment by martinthomson** [comment]
This also works:

```js
window.atob('Zm9vIGJhciBiY  Xo=')
```

**Comment by mnot** [comment]
@phluid61 the problem is that we'd be inventing a new format, one that *isn't* the one we're trying to illustrate.

File: draft-ietf-httpbis-message-signatures.md [hunk 1]
```diff
@@ -143,19 +143,19 @@ For brevity, the term "signature" on its own is used in this document to refer t
 In addition to those listed above, this document uses the following terms:
 
 Decimal String
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:146]:
```suggestion
: 
```

Comment (ioggstream) [reply draft-ietf-httpbis-message-signatures.md:146]: here and elsewhere after ":", imho the extra space improves format: is that correct?

File: draft-ietf-httpbis-message-signatures.md [hunk 2]
```diff
@@ -143,19 +143,19 @@ For brevity, the term "signature" on its own is used in this document to refer t
...
 : An Integer String optionally concatenated with a period "." followed by a second Integer String, representing a positive real number expressed in base 10.  The first Integer String represents the integral portion of the number, while the optional second Integer String represents the fractional portion of the number.  (( Editor's note: There's got to be a definition for this that we can reference. ))
 
 Integer String
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:150]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 3]
```diff
@@ -143,19 +143,19 @@ For brevity, the term "signature" on its own is used in this document to refer t
...
 : A US-ASCII string of one or more digits "0-9", representing a positive integer in base 10. (( Editor's note: There's got to be a definition for this that we can reference. ))
 
 Signer
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:154]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 4]
```diff
@@ -143,19 +143,19 @@ For brevity, the term "signature" on its own is used in this document to refer t
...
 : The entity that is generating or has generated an HTTP Message Signature.
 
 Verifier
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:158]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 5]
```diff
@@ -494,20 +494,20 @@ The `Signature-Input` HTTP header field is a Dictionary Structured Header {{Stru
 The parameters on each `Signature-Input` member value contain metadata about the signature. Each parameter name MUST be a parameter name registered in the IANA HTTP Signatures Metadata Parameters Registry defined in {{param-registry}} of this document. This document defines the following parameters, and registers them as the initial contents of the registry:
 
 alg
-: 
-: RECOMMENDED. The `alg` parameter is a Token containing the name of the signature's Algorithm, as registered in the HTTP Signature Algorithms Registry defined by this document. Verifiers MUST determine the signature's Algorithm from the `keyId` parameter rather than from `alg`. If `alg` is provided and differs from or is incompatible with the algorithm or key material identified by `keyId` (for example, `alg` has a value of `rsa-sha256` but `keyId` identifies an EdDSA key), then implementations MUST produce an error.
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:497]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 6]
```diff
@@ -494,20 +494,20 @@ The `Signature-Input` HTTP header field is a Dictionary Structured Header {{Stru
...
+: RECOMMENDED. The `alg` parameter is a Token containing the name of the signature's Algorithm, as registered in the HTTP Signature Algorithms Registry defined by this document. Verifiers MUST determine the signature's Algorithm from the `keyid` parameter rather than from `alg`. If `alg` is provided and differs from or is incompatible with the algorithm or key material identified by `keyid` (for example, `alg` has a value of `rsa-sha256` but `keyid` identifies an EdDSA key), then implementations MUST produce an error.
 
 created
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:501]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 7]
```diff
@@ -494,20 +494,20 @@ The `Signature-Input` HTTP header field is a Dictionary Structured Header {{Stru
...
 : RECOMMENDED. The `created` parameter is a Decimal containing the signature's Creation Time, expressed as the canonicalized value of the `*created` content identifier, as defined in Section 2.  If not specified, the signature's Creation Time is undefined.  This parameter is useful when signers are not capable of controlling the Date HTTP Header such as when operating in certain web browser environments.
 
 expires
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:505]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 8]
```diff
@@ -494,20 +494,20 @@ The `Signature-Input` HTTP header field is a Dictionary Structured Header {{Stru
...
 : OPTIONAL. The `expires` parameter is a Decimal containing the signature's Expiration Time, expressed as the canonicalized value of the `*expires` content identifier, as defined in Section 2.  If the signature does not have an Expiration Time, this parameter MUST be omitted.  If not specified, the signature's Expiration Time is undefined.
 
-keyId
-: 
-: REQUIRED. The `keyId` parameter is a String whose value can be used by a verifier to identify and/or obtain the signature's Verification Key Material. Further format and semantics of this value are out of scope for this document.
+keyid
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:509]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 9]
```diff
@@ -552,15 +552,15 @@ This document defines HTTP Signature Algorithms, for which IANA is asked to crea
 ### Registration Template {#iana-hsa-template}
 
 Algorithm Name
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:555]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 10]
```diff
@@ -552,15 +552,15 @@ This document defines HTTP Signature Algorithms, for which IANA is asked to crea
...
 : An identifier for the HTTP Signature Algorithm. The name MUST be an ASCII string consisting only of lower-case characters (`"a"` - `"z"`), digits (`"0"` - `"9"`), and hyphens (`"-"`), and SHOULD NOT exceed 20 characters in length.  The identifier MUST be unique within the context of the registry.
 
 Status
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:559]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 11]
```diff
@@ -552,15 +552,15 @@ This document defines HTTP Signature Algorithms, for which IANA is asked to crea
...
 : A brief text description of the status of the algorithm.  The description MUST begin with one of "Active" or "Deprecated", and MAY provide further context or explanation as to the reason for the status.
 
 Description
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:563]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 12]
```diff
@@ -570,16 +570,16 @@ Description
 #### hs2019
 
 Algorithm Name
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:573]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 13]
```diff
@@ -570,16 +570,16 @@ Description
...
 : `hs2019`
 
 Status
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:577]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 14]
```diff
@@ -570,16 +570,16 @@ Description
...
 : active
 
 Description
-: 
-: Derived from metadata associated with keyId. Recommend support for:
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:581]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 15]
```diff
@@ -588,54 +588,54 @@ Description
 
 #### rsa-sha1
 Algorithm Name
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:591]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 16]
```diff
@@ -588,54 +588,54 @@ Description
...
 : `rsa-sha1`
 
 Status
-: 
-: Deprecated; SHA-1 not secure. 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:595]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 17]
```diff
@@ -588,54 +588,54 @@ Description
...
+: Deprecated; SHA-1 not secure.
 
 Description
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:599]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 18]
```diff
@@ -588,54 +588,54 @@ Description
...
 : RSASSA-PKCS1-v1_5 [RFC8017] using SHA-1 [RFC6234]
 
 #### rsa-sha256
 Algorithm Name
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:604]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 19]
```diff
@@ -588,54 +588,54 @@ Description
...
 : `rsa-sha256`
 
 Status
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:608]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 20]
```diff
@@ -588,54 +588,54 @@ Description
...
 : Deprecated; specifying signature algorithm enables attack vector.
 
 Description
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:612]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 21]
```diff
@@ -588,54 +588,54 @@ Description
...
 : RSASSA-PKCS1-v1_5 [RFC8017] using SHA-256 [RFC6234]
 
 #### hmac-sha256
 Algorithm Name
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:617]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 22]
```diff
@@ -588,54 +588,54 @@ Description
...
 : `hmac-sha256`
 
 Status
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:621]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 23]
```diff
@@ -588,54 +588,54 @@ Description
...
 : Deprecated; specifying signature algorithm enables attack vector.
 
 Description
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:625]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 24]
```diff
@@ -588,54 +588,54 @@ Description
...
 : HMAC [RFC2104] using SHA-256 [RFC6234]
 
 #### ecdsa-sha256
 Algorithm Name
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:630]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 25]
```diff
@@ -588,54 +588,54 @@ Description
...
 : `ecdsa-sha256`
 
 Status
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:634]:
```suggestion
: 
```

File: draft-ietf-httpbis-message-signatures.md [hunk 26]
```diff
@@ -588,54 +588,54 @@ Description
...
 : Deprecated; specifying signature algorithm enables attack vector.
 
 Description
-: 
+:
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:638]:
```suggestion
: 
```

**Review by ioggstream — CHANGES_REQUESTED** [review]
recover missing spaces after ":" :P

File: draft-ietf-httpbis-message-signatures.md (hunk 1 above)
Comment (mnot) [reply draft-ietf-httpbis-message-signatures.md:146]: What do you think those spaces do?

Comment (ioggstream) [reply draft-ietf-httpbis-message-signatures.md:146]: With space, the HTML/TXT renders like

```
Signer
    The entity that is generating or has generated an HTTP Message Signature.
```

without space, it renders like

```
Signer
:    The entity that is generating or has generated an HTTP Message Signature.
```

Comment (mnot) [reply draft-ietf-httpbis-message-signatures.md:146]: The correct way to do this is:

~~~
{: vspace="0"}
word:
: definition
~~~

File: draft-ietf-httpbis-message-signatures.md [hunk 27]
```diff
@@ -183,14 +180,14 @@ An HTTP header field value is canonicalized as follows:
 
 This section contains non-normative examples of canonicalized values for header fields, given the following example HTTP message:
 
-~~~
+~~~ http-message
 HTTP/1.1 200 OK
 Server: www.example.com
 Date: Tue, 07 Jun 2014 20:51:35 GMT
-X-OWS-Header:   Leading and trailing whitespace.
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:190]: It seems this WS is intended...

File: draft-ietf-httpbis-message-signatures.md [hunk 28]
```diff
@@ -183,14 +180,14 @@ An HTTP header field value is canonicalized as follows:
...
 This section contains non-normative examples of canonicalized values for header fields, given the following example HTTP message:
 
-~~~
+~~~ http-message
 HTTP/1.1 200 OK
 Server: www.example.com
 Date: Tue, 07 Jun 2014 20:51:35 GMT
-X-OWS-Header:   Leading and trailing whitespace.   
-X-Obs-Fold-Header: Obsolete  
+X-OWS-Header:   Leading and trailing whitespace.
```
Comment (ioggstream) [inline draft-ietf-httpbis-message-signatures.md:187]:
```suggestion
X-OWS-Header:   Leading and trailing whitespace.    
```

@richanna @mnot are those whitespaces needed? The reader won't probably see them :)

**Comment by phluid61** [comment]
Today I learned about RFC 8792.

File: draft-ietf-httpbis-message-signatures.md (hunk 28 above)
Comment (richanna) [reply draft-ietf-httpbis-message-signatures.md:188]: The trailing whitespace on `X-OWS-Header` is important, as it is intended to demonstrate that trailing whitespace is stripped during canonicalization of a header field. I would think trailing whitespace in a codeblock should be preserved. What processor is stripping them out? `kramdown-rfc2629` or `xml2rfc`?

Comment (mnot) [reply draft-ietf-httpbis-message-signatures.md:188]: Oh, no that's just EditorConfig. Do you think people will actually check for trailing whitespace in the RFC?

Comment (richanna) [reply draft-ietf-httpbis-message-signatures.md:192]: If someone copies/pastes the example, they should get what the example is intending to demonstrate. Maybe not super critical in this non-normative example, but I'd like to add a set of canonicalization test cases at some point, where maintaining that trailing whitespace will be very important.

File: sf.json [hunk 29]
```diff
@@ -3,6 +3,12 @@
     "proxy-status": "list",
     "variant-key": "list",
     "variants": "dict",
+    "signature": "dict",
+    "signature-input": "dict",
+    "x-dictionary": "dict",
+    "x-list": "list",
+    "x-list-a": "list",
+    "x-list-b": "list"
```
Comment (mnot) [inline sf.json:11]:
```suggestion
    "x-list-b": "list",
```

File: draft-ietf-httpbis-message-signatures.md (hunk 28 above)
Comment (mnot) [reply draft-ietf-httpbis-message-signatures.md:192]: OK, cool. Will restore and remember not to edit this file with editorconfig enabled; perhaps we should exempt it in `.editorconfig`.

**Review by richanna — APPROVED** [review]

File: draft-ietf-httpbis-message-signatures.md [hunk 30]
```diff
@@ -520,30 +511,34 @@ The `Signature` HTTP header field is a Dictionary Structured Header {{Structured
...
+    CaB8X/I5/+HLZLGvDiezqi6/7p2Gngf5hwZ0lSdy39vyNMaaAT0tKo6nuVw0S1MVg\
+    1Q7MpWYZs0soHjttq0uLIA3DIbQfLiIvK6/l0BdWTU7+2uQj7lBkQAsFZHoA96ZZg\
     FquQrXRlmYOh+Hx5D9fJkXcXe5tmAg==:
 ~~~
 
 Since `Signature-Input` and `Signature` are both defined as Dictionary Structured Headers, they can be used to easily include multiple signatures within the same HTTP message. For example, a signer may include multiple signatures signing the same content with different keys and/or algorithms to support verifiers with different capabilities, or a reverse proxy may include information about the client in header fields when forwarding the request to a service host, and may also include a signature over those fields and the client's signature. The following is a non-normative example of header fields a reverse proxy might add to a forwarded request that contains the signature in the above example:
 
-~~~
+~~~ http-message
+# NOTE: '\' line wrapping per RFC 8792
```
Comment (reschke) [inline draft-ietf-httpbis-message-signatures.md:531]: The string "# NOTE..." is *not* part of the message; it should be outside the artwork.

Comment (mnot) [reply draft-ietf-httpbis-message-signatures.md:531]: Not according to RFC8792.
